File: sentimental-credit/credit.py
from cs50 import get_string
import math

# input card number
cardNumber = get_string("Number: ")

cardNumberlength = len(cardNumber)

# use slicing to store first two digits in variable, start/end/step
firstDigit = cardNumber[0:2]

# this slicing syntax starts -2 from 0, and steps -2 (every other) through the list
multDig = cardNumber[-2::-2]
addDig = cardNumber[-1::-2]

multSum = 0

# the mathfloor method rounds down
for i in range(len(multDig)):
    multSum += math.floor((int(multDig[i]) * 2 / 10)) + math.floor((int(multDig[i]) * 2 % 10))

addSum = 0

for i in range(len(addDig)):
    addSum += int(addDig[i])

checkSum = addSum + multSum
print(firstDigit)

if (cardNumberlength < 13) or (cardNumberlength > 16):
    print("INVALID")
elif (checkSum % 10 == 0) and (firstDigit == "34" or firstDigit == "37"):
    print("AMEX")
elif (checkSum % 10 == 0) and (firstDigit[0] == "4"):
    print("VISA")
elif (checkSum % 10 == 0) and (firstDigit == "51" or firstDigit == "52" or firstDigit == "53" or firstDigit == "54" or firstDigit == "55"):
    print("MASTERCARD")
else:
    print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int lengthFunction(double n, int *firstDigit);
long multSum(long x, long y);
long addSum(long x, long y);

int main(void)
{
    double input;
    int cardNumberlength;
    long sum;
    int firstDigit;

    input = get_long("Number: ");

    /* First card number length and first digits are tabulated*/
    cardNumberlength = lengthFunction(input, &firstDigit);

    /* multSum calculates the multiplied portions, addSum the add-only*/
    sum = multSum(input, cardNumberlength) + addSum(input, cardNumberlength);


    /*Now checksum can be performed, screening for appropriate range and respective card start number(s)*/
    if (cardNumberlength < 13 || cardNumberlength > 16)
    {
        printf("INVALID\n");
    }

    else if (sum % 10 == 0 && (firstDigit == 34 || firstDigit == 37))
    {
        printf("AMEX\n");
    }

    else if (sum % 10 == 0 && (firstDigit / 10 == 4))
    {
        printf("VISA\n");
    }

    else if (sum % 10 == 0 && (firstDigit == 51 || firstDigit == 52 || firstDigit == 53 || firstDigit == 54 || firstDigit == 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

/* Calculates length of number string looping by 10; passes firstDigit by reference*/
int lengthFunction(double n, int *firstDigit)
{
    int i;

    for (i = 0; n > 1; i++)
    {
        n /= 10;
    }
    *firstDigit = n * 100;
    return i;
}

/*Loop uses input value*/
long multSum(long x, long y)
{
    int i = y;
    int multDig[i];
    int multSum = 0;

    for (i = y; x > 0; i--)
    {
        x /= 10;
        multDig[i] += x % 10;
        x /= 10;
        multSum += ((multDig[i] * 2) / 10) + ((multDig[i] * 2) % 10);
    }
    return multSum;
}

/*Loop uses length value*/
long addSum(long x, long y)
{
    int i = y;
    int addDig[i];
    int addSum = 0;

    for (i = y / 2 + 1; i > 0; i--)
    {
        addDig[i] = x % 10;
        x /= 100;
        addSum += addDig[i];
    }
    return addSum;
}